### PR TITLE
better handling of parsed

### LINKED ISSUE #22787: Failed task exposes warning (Removed unexpected internal key)
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

Ansible core

##### ANSIBLE VERSION

v2.3

##### SUMMARY

Since some time, a task failure (caused by an exception) always exposes this (new) warning:
```
 [WARNING]: Removed unexpected internal key in module return: _ansible_parsed = False
```
To me this offers absolutely no additional information to the failure and only adds to the confusion. (As it's the first thing the user sees before seeing the error, and the pink color stands out more than dark red.

I traced this warning to:
https://github.com/ansible/ansible/blame/dd8d699981249038b6d9d8dda05ea57239a7d2d4/lib/ansible/plugins/action/__init__.py#L767

I would prefer to get rid of this warning altogether.

##### STEPS TO REPRODUCE

Make it fail, you will see :-)

### PR DESCRIPTION
**Pull request by bcoca**
##### SUMMARY

fixes #22787

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

core
##### ANSIBLE VERSION

```
2.3
```

### CODE REVIEW

**Comment by dagwieers** [comment]
Thanks !

**Comment by mattclay** [comment]
CI unit test `plugins.action.test_action.TestActionBaseParseReturnedData.test_json_empty` failure:

```
>       self.assertEqual(len(res), 0)
E       AssertionError: 1 != 0

test/units/plugins/action/test_action.py:594: AssertionError
```

File: lib/ansible/plugins/action/__init__.py [hunk 1]
```diff
@@ -763,7 +762,7 @@ def _execute_module(self, module_name=None, module_args=None, tmp=None, task_var
 
     def _remove_internal_keys(self, data):
         for key in list(data.keys()):
-            if key.startswith('_ansible_') or key in C.INTERNAL_RESULT_KEYS:
+            if key.startswith('_ansible_') and key != '_ansible_parsed' or key in C.INTERNAL_RESULT_KEYS:
```
Comment (alikins) [inline lib/ansible/plugins/action/__init__.py:769]: Seems reasonable for a 2.3 targeted fix.

Long term (after 2.3) may be useful to add a constant for the keys that are whitelisted ( ie something like `and key not in C.INTERNAL_RESULT_KEYS_WHITELIST` )

**Comment by alikins** [comment]
re plugins.action.test_action.TestActionBaseParseReturnedData.test_json_empty failure.

Looks like that test should be updated to expect _parse_returned_data() to always return a dict with 1 key on success (the _ansible_parsed item). 

Only concern would be if anything relies on truthiness of _parse_returned_data() return, but it doesn't look like anything does.

**Review by alikins — COMMENTED** [review]
Approve once test case is updated.

File: lib/ansible/plugins/action/__init__.py (hunk 1 above)
Comment (bcoca) [reply lib/ansible/plugins/action/__init__.py:769]: well, parsed should be only key needed at this point, anything else should be consumed inside the previous code or generated afterwards

**Pull request merged**